passes/sat/qbfsat.cc

```diff
@@ -0,0 +1,522 @@
+	log("                 \\____ $$$|__/|________/|__/|_______/|__/\n");
+	log("                       \\__/                              \n");
+	log("\n");
+}
+
+struct QbfSatPass : public Pass {
+	QbfSatPass() : Pass("qbfsat", "solve a 2QBF-SAT problem in the circuit") { }
+	void help() YS_OVERRIDE
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    qbfsat [options] [selection]\n");
+		log("\n");
+		log("This command solves a 2QBF-SAT problem defined over the currently selected module.\n");
+		log("Existentially-quantified variables are declared by assigning a wire \"$anyconst\".\n");
+		log("Universally-quantified variables may be explicitly declared by assigning a wire\n");
+		log("\"$allconst\", but module inputs will be treated as universally-quantified variables\n");
+		log("by default.\n");
+		log("\n");
+		log("    -nocleanup\n");
+		log("        Do not delete temporary files and directories.  Useful for\n");
+		log("        debugging.\n");
+		log("\n");
+		log("    -dump-final-smt2 <file>\n");
+		log("        Pass the --dump-smt2 option to yosys-smtbmc.\n");
+		log("\n");
+		log("    -assume-outputs\n");
+		log("        Add an $assume cell for the conjunction of all one-bit module output wires.\n");
+		log("\n");
+		log("    -sat\n");
+		log("        Generate an error if the solver does not return \"sat\".\n");
+		log("\n");
+		log("    -unsat\n");
+		log("        Generate an error if the solver does not return \"unsat\".\n");
+		log("\n");
+		log("    -specialize\n");
+		log("        Replace all \"$anyconst\" cells with constant values determined by the solver.\n");
+		log("\n");
+		log("    -specialize-from-file <solution file>\n");
+		log("        Do not run the solver, but instead only attempt to replace all \"$anyconst\"\n");
+		log("        cells in the current module with values provided by the specified file.\n");
+		log("\n");
+		log("    -write-solution <solution file>\n");
+		log("        Write the assignments discovered by the solver for all \"$anyconst\" cells\n");
+		log("        to the specified file.");
+		log("\n");
+		log("\n");
+	}
+	void execute(std::vector<std::string> args, RTLIL::Design *design) YS_OVERRIDE
+	{
+		log_header(design, "Executing QBF-SAT pass (solving QBF-SAT problems in the circuit).\n");
+		QbfSolveOptions opt = parse_args(args);
+		extra_args(args, opt.argidx, design);
+
+
+		RTLIL::Module *module = NULL;
+		for (auto mod : design->selected_modules()) {
+			if (module)
+				log_cmd_error("Only one module must be selected for the QBF-SAT pass! (selected: %s and %s)\n", log_id(module), log_id(mod));
+			module = mod;
+		}
+		if (module == NULL)
+			log_cmd_error("Can't perform QBF-SAT on an empty selection!\n");
+
+		if (!opt.specialize_from_file) {
+			//Save the design to restore after modiyfing the current module.
+			std::string module_name = module->name.str();
+			Pass::call(design, "design -save _qbfsat_tmp");
```

Comment:
See `CodingReadme` section of ScriptPass: A pass should, if possible, either do proper work, or call other passes, but not mix both, because that would make things really hard to debug. Not 100% sure if this would qualify for one of the few rare exceptions for this rule, I guess that depends on how complex it would be to refactor this in that fashion.

Also: This seems to be a good use case for `design -push` and `design -pop`.

Reply:
> because that would make things really hard to debug.

Why does it make things difficult to debug?

> Also: This seems to be a good use case for design -push and design -pop

It would be, but unfortunately `-push` also clears the design, and there is not a `-duplicate` option for `design` (i.e. a way to push the current design onto the design stack without clearing the current design).  So the only way is to use `-save` or `-stash` with an explicit name like `_qbfsat_tmp`.  Perhaps I will add a `-duplicate` option to `design` and use that; I agree it's inelegant to use this explicit named save/load hack.